pkg/pubsub/pubsub_client.go

```diff
@@ -0,0 +1,123 @@
+package pubsub
+
+import (
+	"context"
+	"flag"
+	"fmt"
+	"strings"
+
+	"github.com/open-policy-agent/gatekeeper/pkg/pubsub/dapr"
+	"github.com/open-policy-agent/gatekeeper/pkg/pubsub/rabbitmq"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+func init() {
+	flag.Var(pubSubs, "pub-sub-tool", "Preferred pub-sub tool, e.g dapr, rabbitmq. This flag can be declared more than once. Omitting will default to not using any pub sub tool.")
+}
+
+var pubSubs = newPubSubSet(map[string]PubSub{
+	dapr.Name:     &dapr.Dapr{},
+	rabbitmq.Name: &rabbitmq.RabbitMQ{},
+},
+)
+
+// PubSub is the interface that wraps pubsub methods.
+type PubSub interface {
+	// Publish single message over a specific topic/channel
+	Publish(data interface{}, topic string) error
+
+	// Publish a batch of messages over a specific topic/channel
+	PublishBatch(data interface{}, topic string) error
+
+	// Initiate the pub sub client
+	NewClient(ctx context.Context, client client.Client) error
+
+	// Get the name of pub sub tool
+	GetName() string
+
+	// Determines if publish batching is enabled or not
+	IsBatchingEnabled() bool
```

Comment:
For base purposes, I think these methods make sense. I am coming short of coming up with additional methods that we might need from this interface. Please suggest additional methods if anything comes to mind, else we can go ahead with this for now and add something in the future as well. Hopefully, it won't be a big change if we are not writing drivers for a lot of pub-sub tools.

Reply:
They largely make sense, the biggest concern is the constructor